codegen/smithy-go-codegen/src/main/java/software/amazon/smithy/go/codegen/integration/IdempotencyTokenMiddlewareGenerator.java

```diff
@@ -34,37 +36,66 @@
 import software.amazon.smithy.model.shapes.Shape;
 import software.amazon.smithy.model.shapes.ShapeId;
 import software.amazon.smithy.model.traits.IdempotencyTokenTrait;
-import software.amazon.smithy.utils.ListUtils;
 
 public class IdempotencyTokenMiddlewareGenerator implements GoIntegration {
+    List<RuntimeClientPlugin> runtimeClientPlugins = new ArrayList<>();
+
     private void execute(
+            Model model,
             GoWriter writer,
             SymbolProvider symbolProvider,
             OperationShape operation,
-            MemberShape idempotencyTokenMemberShape,
-            Symbol inputSymbol
+            MemberShape idempotencyTokenMemberShape
     ) {
         GoStackStepMiddlewareGenerator middlewareGenerator =
-            GoStackStepMiddlewareGenerator.createInitializeStepMiddleware(
-                    getIdempotencyTokenMiddlewareName(operation));
+                GoStackStepMiddlewareGenerator.createInitializeStepMiddleware(
+                        getIdempotencyTokenMiddlewareName(operation));
 
+        Shape inputShape = model.expectShape(operation.getInput().get());
+        Symbol inputSymbol = symbolProvider.toSymbol(inputShape);
         String memberName = symbolProvider.toMemberName(idempotencyTokenMemberShape);
-            middlewareGenerator.writeMiddleware(writer, (generator, middlewareWriter) -> {
+        middlewareGenerator.writeMiddleware(writer, (generator, middlewareWriter) -> {
             middlewareWriter.write("input, ok := in.Parameters.($P)", inputSymbol);
             middlewareWriter.write("if !ok { return out, metadata, "
                     + "fmt.Errorf(\"expected middleware input to be of type $P \")}", inputSymbol);
             middlewareWriter.addUseImports(SmithyGoDependency.FMT);
 
             middlewareWriter.openBlock("if input.$L == nil {", "}", memberName, () -> {
-            writer.addUseImports(SmithyGoDependency.SMITHY_RAND);
-            writer.addUseImports(SmithyGoDependency.CRYPTORAND);
-            writer.write("uuid := smithyrand.NewUUID(randReader)");
-            writer.write("v, err := uuid.GetUUID()");
-            writer.write("if err != nil { return out, metadata, err}");
-                middlewareWriter.write("input.$L = &v", memberName);
+                middlewareWriter.write("t, err := m.tokenProvider.GetToken()");
+                middlewareWriter.write(" if err != nil { return out, metadata, err }");
+                middlewareWriter.write("input.$L = &t", memberName);
```

Comment:
if token provider is nil, probably want to skip this middleware, and delegate to `next`

Reply:
Updated to add nil check